tests/test_daemon.cpp

```diff
@@ -860,13 +870,48 @@ TEST_P(LaunchImgSizeSuite, launches_with_correct_disk_size)
         send_command(all_parameters, trash_stream, stream);
         EXPECT_THAT(stream.str(), AllOf(HasSubstr("Requested disk"), HasSubstr("below minimum for this image")));
     }
+    else if (other_command_line_parameters.size() > 0 &&
+             mp::MemorySize(other_command_line_parameters[1]) > mp::MemorySize(available_bytes_str + "B"))
+    {
+        std::stringstream stream;
+        EXPECT_CALL(*mock_factory, create_virtual_machine(_, _)).Times(0);
+        send_command(all_parameters, trash_stream, stream);
+        EXPECT_THAT(stream.str(), AllOf(HasSubstr("Available disk"), HasSubstr("below requested/default size")));
+    }
     else
     {
         EXPECT_CALL(*mock_factory, create_virtual_machine(_, _));
         send_command(all_parameters);
     }
 }
 
+TEST_P(InvalidDataDirectorySuite, launch_fails_with_invalid_data_directory)
+{
+    auto mock_factory = use_a_mock_vm_factory();
+    const auto param = GetParam();
+    const auto& first_command_line_parameter = std::get<0>(param);
+    const auto& other_command_line_parameters = std::get<1>(param);
+    const auto& img_size_str = std::get<2>(param);
+
+    auto mock_image_vault = std::make_unique<NiceMock<mpt::MockVMImageVault>>();
+    ON_CALL(*mock_image_vault.get(), minimum_image_size_for(_)).WillByDefault([&img_size_str](auto...) {
+        return mp::MemorySize{img_size_str};
+    });
+
+    config_builder.vault = std::move(mock_image_vault);
+    config_builder.data_directory = QString("invalid_data_directory");
+    mp::Daemon daemon{config_builder.build()};
+
+    std::vector<std::string> all_parameters{first_command_line_parameter};
+    for (const auto& p : other_command_line_parameters)
+        all_parameters.push_back(p);
+
+    std::stringstream stream;
+    EXPECT_CALL(*mock_factory, create_virtual_machine(_, _)).Times(0);
+    send_command(all_parameters, trash_stream, stream);
+    EXPECT_THAT(stream.str(), HasSubstr("Failed to determine information about the volume containing"));
+}
```

Comment:
This test can be a lot simpler than the one you used as inspiration. For instance, you don't need all the parameters, since you use only one value; then you don't need to construct a complex command either; and neither do you need all the mocking. Consider this:

```c++
TEST_P(InvalidDataDirectorySuite, launch_fails_with_invalid_data_directory)
{
    auto mock_factory = use_a_mock_vm_factory();
    config_builder.data_directory = QString("invalid_data_directory");
    mp::Daemon daemon{config_builder.build()};

    std::stringstream stream;
    EXPECT_CALL(*mock_factory, create_virtual_machine(_, _)).Times(0);
    send_command({GetParam()}, trash_stream, stream);
    EXPECT_THAT(stream.str(), HasSubstr("Failed to determine information about the volume containing"));
}
```
You'd need to change the param interface to receive only one string. You could then do something very similar for the other test, adding just `--disk ...` to the command.

Reply:
Thank you for this information! I was wondering about simplifying this test, but I didn't want to accidentally remove parts that were actually necessary -- so I appreciate you taking the time to discuss how to simplify it. Now I've made the changes you suggested.